controller_manager:
  ros__parameters:
    update_rate: 10  # Match your sensor frequency
    
    # Controllers to load
    diff_drive_controller:
      type: diff_drive_controller/DiffDriveController
      
    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

# Differential drive controller configuration
diff_drive_controller:
  ros__parameters:
    # Joint names for skid-steer configuration
    left_wheel_names: ["front_left_wheel_joint"]
    right_wheel_names: ["rear_right_wheel_joint"]
    
    # Physical parameters (your measured values)
    wheel_separation: 0.148
    wheel_radius: 0.03
    
    # Limits (based on your motor characterization)
    wheel_separation_multiplier: 1.0
    left_wheel_radius_multiplier: 1.0
    right_wheel_radius_multiplier: 1.0
    
    # Publishing
    publish_rate: 10.0  # Match control frequency
    odom_frame_id: odom
    base_frame_id: base_link
    pose_covariance_diagonal : [0.001, 0.001, 0.001, 0.001, 0.001, 0.01]
    twist_covariance_diagonal: [0.001, 0.001, 0.001, 0.001, 0.001, 0.01]

        # Optional: Remove or increase velocity limits
    linear.x.has_velocity_limits: true
    linear.x.max_velocity: 0.21 # 25 cm/s (even more headroom!)
    angular.z.has_velocity_limits: true  
    angular.z.max_velocity: 2.8  # 3.5 rad/s

# Joint state broadcaster (publishes joint states for RViz)
joint_state_broadcaster:
  ros__parameters:
    joints:
      - front_left_wheel_joint
      - rear_right_wheel_joint
      - front_right_wheel_joint  
      - rear_left_wheel_joint